lib/pbench/test/unit/server/database/__init__.py

```diff
@@ -0,0 +1,351 @@
+"""The SQLAlchemy mock infrastructure here is not completely generalized, but
+is currently sufficient to support the ServerConfig and Audit DB unit tests.
+"""
+
+import operator
+from typing import Optional
+
+from sqlalchemy import Column
+from sqlalchemy.util import immutabledict
+
+from pbench.server.database.database import Database
+
+
+def dumpem(object: object, depth: int = 0, max_depth: int = 20):
+    """
+    Useful recursive function to help debug SQLAlchemy's deep and mostly not
+    documented class hierarchy! NOTE: this can generate an enormous amount of
+    output and should be used carefully. While simple in principle, it's worth
+    archiving for future use.
+
+    The recursion depth is limited to 20 by default. That's arbitrary, but we
+    don't want to overflow.
+
+    Args:
+        object: Any Python object type
+        depth: indentation levels (two spaces)
+        max_depth: maximum recursion depth
+    """
+    if depth > max_depth:
+        return
+    indent = " " * (depth * 2)
+    print(f"{indent}{object} [{type(object).__name__}]")
+    if isinstance(object, list):
+        for e in object:
+            dumpem(e, depth + 1, max_depth)
+    elif isinstance(object, (dict, immutabledict)):
+        for n, v in object.items():
+            print(f"{indent}  '{n}'={v} [{type(v).__name__}]")
+            dumpem(v, depth + 2, max_depth)
+    elif (
+        object is not None
+        and not isinstance(object, (bool, int, float, str))
+        and hasattr(object, "__dict__")
+    ):
+        for n, v in object.__dict__.items():
+            print(f"{indent}  {n}={v} [{type(v).__name__}]")
+            dumpem(v, depth + 2, max_depth)
```

Comment:
I _think_ you can combine these two cases:
```
    elif hasattr(object, "items") or hasattr(object, "__dict__"):
        d = object if hasattr(object, "items") else object.__dict__
        for n, v in d.items():
            print(f"{indent}  '{n}'={v} [{type(v).__name__}]")
            dumpem(v, depth + 2, max_depth)
```
(Then I don't have to comment on the difference in quotation between lines 38 and 46....)

BTW, this just dumps the structure, right?  (The scalars/values aren't printed.)  Should there be a final `else` which just prints the `str()` or `repr()` of the "object"?  (The type hint for `object` implies that it's not a scalar, but line 42 sort of undermines that idea.  🙂)

Reply:
Back at line 32 we dump whatever the current instance is, along with its type. *Iff* it's a collection we can interpret, we then curse and recurse to get its elements. This is already slightly redundant for the builtin Python collections of Python objects, which get adequately stringified through the collection, but I needed more for the value of things like `list[object]` where "object" might be a non-Python object that doesn't adequately (and recursively) stringify itself, so I accepted the redundancy of e.g. `[1] [list]` followed by `  1 [int]` 😆 

I could possibly combine those; not at all sure it's worthwhile.

But, yeah, I should probably use `'{n}'` consistently, or even `{n!r}` where I'm pretty sure I'm working with strings ... although I'm not entirely sure that's ever true here.

Reply:
> Back at line 32 we dump whatever the current instance is, along with its type.

Ah.  I missed that.  And, now that I look, it looks like we would likely have previously printed the value at line 38 or 46 as well.  Nevermind -- twice is way better than none.

> I could possibly combine those; not at all sure it's worthwhile.

Aside from the consistency thing, I was looking at the compound conditional at lines 41-43 and thinking that the only test which matters is the last one, because if any of the previous ones aren't true then `object` won't have a `__dict__` attribute, so you can boil it down to one test.  (And, then I noted that you could separate the iteration from the bodies of the loops, which were (ostensibly) identical.)

Reply:
Looking at the code again, I tried to adopt some of your ideas here to "streamline" but every time I look (and try it out) I think of all sorts of one-plus changes I really don't want to dig into for a silly little debugging utility. (E.g., many of the classes have long multi-line `__doc__` strings, and I think about reformatting them, or conditionally cropping them...)

I'm kinda hoping that what I've got is "good enough" and we can just drop this. And, yeah, I'm fully aware of the irony of that since you already resolved your conversation and I'm the one continuing to reply and mess with the code, but that's not the point. Or maybe it is. 😆